File: app/src/main/java/com/github/mbarrben/moviedb/BaseFragment.java
package com.github.mbarrben.moviedb;

import android.app.Activity;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import butterknife.ButterKnife;

public abstract class BaseFragment extends Fragment {

  @Override
  public void onAttach(Activity activity) {
    super.onAttach(activity);
    injectDependencies();
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    return inflater.inflate(getFragmentLayout(), container, false);
  }

  @Override
  public void onViewCreated(View view, Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    injectViews(view);
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    ButterKnife.reset(this);
  }

  protected abstract int getFragmentLayout();

  private void injectDependencies() {
    ((BaseActivity) getActivity()).inject(this);
  }

  private void injectViews(final View view) {
    ButterKnife.inject(this, view);
  }
}

File: app/src/main/kotlin/com/github/mbarrben/moviedb/BaseFragment.kt
package com.github.mbarrben.moviedb

import android.content.Context
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.mbarrben.moviedb.di.HasComponent

abstract class BaseFragment(@LayoutRes val layoutRes: Int) : Fragment() {

  override fun onAttach(context: Context?) {
    super.onAttach(context)
    inject()
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
      = inflater.inflate(layoutRes, container)

  fun <T> getComponent(componentType: Class<T>): T = componentType.cast((activity as HasComponent<*>).component)

  abstract fun inject()

}
